pkg/topo/topo.go

```diff
@@ -2,70 +2,122 @@
 	"context"
+
+	"github.com/onosproject/onos-ric-sdk-go/pkg/topo/options"
+
+	"github.com/onosproject/onos-ric-sdk-go/pkg/utils/creds"
+	"google.golang.org/grpc/credentials"
+
 	topoapi "github.com/onosproject/onos-api/go/onos/topo"
-	"github.com/onosproject/onos-lib-go/pkg/errors"
+
 	"github.com/onosproject/onos-lib-go/pkg/logging"
-	"google.golang.org/grpc/status"
-	"io"
+
+	"time"
+
+	client "github.com/onosproject/onos-ric-sdk-go/pkg/topo/client"
+
+	"github.com/onosproject/onos-ric-sdk-go/pkg/topo/connection"
+
+	"github.com/onosproject/onos-lib-go/pkg/southbound"
+	"google.golang.org/grpc"
 )
 
-var log = logging.GetLogger("topo", "watch")
+var log = logging.GetLogger("topo")
 
-// Watch provides a simple facility for the application to watch for changes in the topology.
-func (c *topoClient) Watch(ctx context.Context, ch chan<- topoapi.Event, filter ...FilterOption) error {
-	// Issue a watch request with replay and passing through any service-side filter criteria, i.e.
-	// those supported by the topo.Watch method (TBD)
-	stream, err := c.client.Watch(ctx, &topoapi.WatchRequest{})
-	if err != nil {
-		log.Error("Unable to issue topology watch request", err)
-		stat, ok := status.FromError(err)
-		if ok {
-			return errors.FromStatus(stat)
-		}
-		return err
-	}
+// Client is a topo client
+type Client interface {
 
-	// Kick off a go routine that reads events from the topology watch stream and subjects them to additional
-	// client-side filter criteria; objects that pass the filter get put on the application provided channel.
-	go handleWatchStream(ch, stream, filter...)
-	return nil
+	// Update updates a topo object
+	Update(ctx context.Context, object *topoapi.Object) error
+
+	// Get gets a topo object with a given ID
+	Get(ctx context.Context, id topoapi.ID) (*topoapi.Object, error)
+
+	// Watch provides a simple facility for the application to watch for changes in the topology
+	Watch(ctx context.Context, ch chan<- topoapi.Event, opts ...options.Option) error
+
+	// List of topo objects
+	List(ctx context.Context, opts ...options.Option) ([]topoapi.Object, error)
 }
 
-// handleWatchStream reads events from the topology watch stream and subjects them to additional
-// client-side filter criteria; objects that pass the filter get put on the application provided channel.
-func handleWatchStream(ch chan<- topoapi.Event, stream topoapi.Topo_WatchClient, filter ...FilterOption) {
-	defer close(ch)
-	for {
-		resp, err := stream.Recv()
-		if err == io.EOF || err == context.Canceled {
-			break
-		}
+// NewClient creates a new topo client
+func NewClient(opts ...options.Option) (Client, error) {
+	options := options.Options{
+		Service: options.ServiceOptions{
+			Host: options.DefaultServiceHost,
+			Port: options.DefaultServicePort,
+		},
+	}
 
+	for _, opt := range opts {
+		opt.Apply(&options)
+	}
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithStreamInterceptor(southbound.RetryingStreamClientInterceptor(100 * time.Millisecond)),
```

Comment:
We probably need to add the unary retrying interceptor here as well. Also, I’ve recently made some improvements to the interceptors to add options to them in Atomix, as well, and we should probably port those to onos-lib-go and expose them via the options here.

Reply:
Added the one we have in southbound for now until we add new packages in onos-lib-go